(TK-186) Port ConfigNode tests

fpringvaldsen opened the pull request:
Some of the implementation of this is a bit different from what's in master at the time of this writing. This is due to the fact that, when porting these tests, I discovered some issues with the upstream tests that needed to be fixed/cleaned up. This PR contains the tests in the upstream master branch, along with the changes I've made in https://github.com/typesafehub/config/pull/301.

jpinsonault commented on lib/hocon/impl/config_node_comment.rb:
```diff
@@ -0,0 +1,19 @@
+    end
+  end
+
+  def commentText
```
should this be `comment_text `?

jpinsonault commented on lib/hocon/impl/config_node_complex_value.rb:
```diff
@@ -0,0 +1,53 @@
+    @children.each do |child|
+      tokens += child.tokens
+    end
+    tokens
```
Not really important, but this might look cleaner as `token = @children.map { |child| child.tokens }`

fpringvaldsen replied:
@jpinsonault That causes an issue in that it will cause `tokens` to be an array of arrays rather than an array of tokens

jpinsonault replied:
oh you're totally right, my bad. Carry on

jpinsonault commented on lib/hocon/impl/config_node_include.rb:
```diff
@@ -0,0 +1,32 @@
+    @children.each do |child|
+      tokens += child.tokens
+    end
+    tokens
```
ditto the .map comment

jpinsonault commented:
Just a couple really minor comments, otherwise looks great!

fpringvaldsen commented:
@jpinsonault I've fixed that typo.

Also, I went ahead and changed all the classes I implemented that are abstract in the upstream library into modules, since that seems to be the way we're handling that.

jpinsonault commented:
Sounds good!

:+1:
